### PR TITLE
Implement studio url config

### PR DESCRIPTION
**Pull request by awais786**
Implement Studio URL in partner config.

https://openedx.atlassian.net/browse/ECOM-6932

### CODE REVIEW

**Comment by awais786** [comment]
@bderusha Please review.

File: course_discovery/apps/core/admin.py [hunk 1]
```diff
@@ -56,6 +56,10 @@ class PartnerAdmin(admin.ModelAdmin):
             'fields': ('marketing_site_url_root', 'marketing_site_api_url', 'marketing_site_api_username',
                        'marketing_site_api_password',)
         }),
+        (_('Miscellaneous Configuration'), {
```
Comment (bderusha) [inline course_discovery/apps/core/admin.py]: LMS/Studio Configuration

File: course_discovery/apps/publisher/models.py [hunk 2]
```diff
@@ -307,6 +313,14 @@ def post_back_url(self):
     def created_by(self):
         return self.history.order_by('history_date').first().history_user     # pylint: disable=no-member
 
+    @property
+    def course_studio_url(self):
```
Comment (bderusha) [inline course_discovery/apps/publisher/models.py]: just `studio_url`

File: course_discovery/templates/publisher/course_detail/_widgets.html [hunk 3]
```diff
@@ -30,7 +30,7 @@ <h5 class="hd-5 emphasized course-runs-heading">{% trans "COURSE RUNS" %}</h5>
                     <div class="studio-url">
                         <span class="studio-url-heading">{% trans "STUDIO URL" %} - </span>
                         {% if course_run.lms_course_id %}
-                            <a class="studio-link" href="#" target="_blank">{{ course_run.lms_course_id }}</a>
+                            <a class="studio-link" href="{% if course_run.course_studio_url %}{{ course_run.course_studio_url }}{% endif %}" target="_blank">{{ course_run.lms_course_id }}</a>
```
Comment (bderusha) [inline course_discovery/templates/publisher/course_detail/_widgets.html]: I dont think we need an `if block` here.

**Review by bderusha — APPROVED** [review]
👍 once comments addressed

**Comment by awais786** [comment]
@waheedahmed @tasawernawaz Please review.

File: course_discovery/apps/publisher/models.py [hunk 4]
```diff
@@ -307,6 +313,14 @@ def post_back_url(self):
     def created_by(self):
         return self.history.order_by('history_date').first().history_user     # pylint: disable=no-member
 
+    @property
+    def studio_url(self):
+        if hasattr(self.course, 'partner') and self.lms_course_id:
```
Comment (waheedahmed) [inline course_discovery/apps/publisher/models.py]: No need to use `hasattr()`, you can simplify this condition `if self.course.partner and self.lms_course_id:`

**Review by waheedahmed — APPROVED** [review]
One minor comment otherwise 👍

**Comment by waheedahmed** [comment]
You also need to add `studio_url` here [add_update_course_form.html#L202](https://github.com/edx/course-discovery/blob/master/course_discovery/templates/publisher/add_update_course_form.html#L202)

File: course_discovery/templates/publisher/course_detail/_widgets.html [hunk 5]
```diff
@@ -30,7 +30,7 @@ <h5 class="hd-5 emphasized course-runs-heading">{% trans "COURSE RUNS" %}</h5>
                     <div class="studio-url">
                         <span class="studio-url-heading">{% trans "STUDIO URL" %} - </span>
                         {% if course_run.lms_course_id %}
```
Comment (waheedahmed) [inline course_discovery/templates/publisher/course_detail/_widgets.html]: Update this check also with `{% if course_run.studio_url %}`.

File: course_discovery/templates/publisher/add_update_course_form.html [hunk 6]
```diff
@@ -198,8 +198,8 @@ <h2 class="layout-title">{% trans "Studio Instance" %}</h2>
                             {% else %}
                                 <div class="studio-url">
                                     <span class="studio-url-heading">{% trans "STUDIO URL" %} - </span>
-                                    {% if run_form.lms_course_id.value %}
-                                        <a class="studio-link" href="#" target="_blank">{{ course_run.lms_course_id }}</a>
+                                    {% if course_run.lms_course_id and course_run.studio_url %}
```
Comment (waheedahmed) [inline course_discovery/templates/publisher/add_update_course_form.html]: No need for this `course_run.lms_course_id` here, you are already doing this in `studio_url` property.

File: course_discovery/templates/publisher/course_detail/_widgets.html [hunk 7]
```diff
@@ -29,8 +29,8 @@ <h5 class="hd-5 emphasized course-runs-heading">{% trans "COURSE RUNS" %}</h5>
                     </div>
                     <div class="studio-url">
                         <span class="studio-url-heading">{% trans "STUDIO URL" %} - </span>
-                        {% if course_run.lms_course_id %}
-                            <a class="studio-link" href="#" target="_blank">{{ course_run.lms_course_id }}</a>
+                        {% if course_run.lms_course_id and course_run.studio_url %}
```
Comment (waheedahmed) [inline course_discovery/templates/publisher/course_detail/_widgets.html]: No need for this `course_run.lms_course_id` here, you are already doing this in `studio_url` property.

**Comment by awais786** [comment]
@bderusha by default migration created is 0002_partner_studio_url. The app also contains squashed migration. So 0002 is appearing twice but with different label.

**Comment by awais786** [comment]
@clintonb I have added a new field in partners table. And by default new migration created as 0002_partner_studio_url. This app has squashed migration. Just want to confirm this sequence is fine ?

File: course_discovery/apps/core/admin.py [hunk 8]
```diff
@@ -56,6 +56,10 @@ class PartnerAdmin(admin.ModelAdmin):
             'fields': ('marketing_site_url_root', 'marketing_site_api_url', 'marketing_site_api_username',
                        'marketing_site_api_password',)
         }),
+        (_('LMS/Studio Configuration'), {
+            'description': _('Configurations used for other services.'),
```
Comment (clintonb) [inline course_discovery/apps/core/admin.py]: This description conflicts with the fieldset name. You can always just add this to the first fieldset (with no label), and section it off later.

File: course_discovery/apps/publisher/models.py [hunk 9]
```diff
@@ -307,6 +313,14 @@ def post_back_url(self):
     def created_by(self):
         return self.history.order_by('history_date').first().history_user     # pylint: disable=no-member
 
+    @property
+    def studio_url(self):
+        if self.lms_course_id and self.course.partner and self.course.partner.studio_url:
```
Comment (clintonb) [inline course_discovery/apps/publisher/models.py]: Under what conditions would a course not have a partner? How do we ensure this condition is never met?

File: course_discovery/apps/publisher/tests/test_model.py [hunk 10]
```diff
@@ -6,6 +6,7 @@
 from django_fsm import TransitionNotAllowed
 from guardian.shortcuts import assign_perm
 
+from course_discovery.apps.course_metadata.tests.factories import OrganizationFactory
```
Comment (clintonb) [inline course_discovery/apps/publisher/tests/test_model.py]: Alphabetize.

**Review by clintonb — COMMENTED** [review]
Yes, your migrations are correct.

File: course_discovery/apps/publisher/models.py (hunk 9 above)
Comment (awais786) [reply course_discovery/apps/publisher/models.py]: I think course will always have a partner. It's better to remove the partner condition from here.

**Comment by awais786** [comment]
@clintonb You can review this again.

File: course_discovery/apps/core/tests/factories.py [hunk 11]
```diff
@@ -36,6 +36,7 @@ class PartnerFactory(factory.DjangoModelFactory):
     oidc_url_root = '{root}'.format(root=FuzzyUrlRoot().fuzz())
     oidc_key = FuzzyText().fuzz()
     oidc_secret = FuzzyText().fuzz()
+    studio_url = '{root}/'.format(root=FuzzyUrlRoot().fuzz())
```
Comment (clintonb) [inline course_discovery/apps/core/tests/factories.py]: Drop the `format` and keep this simple.

File: course_discovery/apps/publisher/models.py [hunk 12]
```diff
@@ -307,6 +313,14 @@ def post_back_url(self):
     def created_by(self):
         return self.history.order_by('history_date').first().history_user     # pylint: disable=no-member
 
+    @property
+    def studio_url(self):
+        if self.lms_course_id and self.course.partner.studio_url:
+            path = 'course/{lms_course_id}'.format(lms_course_id=self.lms_course_id)
+            return urljoin(self.course.partner.studio_url, path)
```
Comment (clintonb) [inline course_discovery/apps/publisher/models.py:320]: Make sure this works regardleas of whether the Studio URL has a trailing slash.

**Review by clintonb — APPROVED** [review]

File: course_discovery/apps/publisher/models.py (hunk 12 above)
Comment (awais786) [reply course_discovery/apps/publisher/models.py:320]: Verified.

**Pull request merged**